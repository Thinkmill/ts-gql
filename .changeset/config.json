{
  "$schema": "https://unpkg.com/@changesets/config@1.0.3/schema.json",
  "changelog": [
    "@changesets/changelog-github",
    {
      "repo": "Thinkmill/ts-gql"
    }
  ],
  "commit": false,
  "linked": [],
  "access": "public",
  "baseBranch": "main"
}
