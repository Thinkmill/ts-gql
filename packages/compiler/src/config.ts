import path from "path";
import { findPkgJsonFieldUp } from "find-pkg-json-field-up";
import { getSchemaFromOptions } from "./get-schema";

export async function getRawConfig(cwd: string) {
  let { packageJson, directory } = await findPkgJsonFieldUp("ts-gql", cwd);
  let field = packageJson["ts-gql"];
  if (
    typeof field === "object" &&
    field !== null &&
    typeof field.schema === "string"
  ) {
    return {
      schema: path.resolve(directory, field.schema),
      directory,
    };
  }
  throw new Error("Config not found");
}

export async function getConfig(cwd: string) {
  let config = await getRawConfig(cwd);
  return {
    directory: config.directory,
    schema: await getSchemaFromOptions(config.schema),
  };
}
