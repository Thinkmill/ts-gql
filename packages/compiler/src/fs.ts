import gracefulFS from "graceful-fs";
import { promisify } from "util";

export const writeFile = promisify(gracefulFS.writeFile);
export const readFile = promisify(gracefulFS.readFile);
export const mkdir = promisify(gracefulFS.mkdir);
export const readdir = promisify(gracefulFS.readdir);
export const unlink = promisify(gracefulFS.unlink);
