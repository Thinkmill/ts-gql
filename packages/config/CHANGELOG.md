# @ts-gql/config

## 0.5.0

### Minor Changes

- [`fba7341`](https://github.com/Thinkmill/ts-gql/commit/fba7341a1418e0a9d555172dc5c6e86899fa6ed3) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Add `readonlyTypes` option

## 0.4.0

### Minor Changes

- [`4be8faa`](https://github.com/Thinkmill/ts-gql/commit/4be8faafa0fba17efa491a0aec8ddbb472aa5572) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Replace `nonOptionalTypename` option with `addTypename` option

## 0.3.0

### Minor Changes

- [`d798897`](https://github.com/Thinkmill/ts-gql/commit/d7988972e801c41bb96aaa4dec5763ebae73e30e) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Add nonOptionalTypename option

## 0.2.0

### Minor Changes

- [`caa1974`](https://github.com/Thinkmill/ts-gql/commit/caa19743de1aa1345795691b8d4eea58c052fc8f) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Support scalars option

## 0.1.0

### Minor Changes

- [`e42383b`](https://github.com/Thinkmill/ts-gql/commit/e42383b5970a554462384f9851aabc173f7fcf52) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Initial Release
