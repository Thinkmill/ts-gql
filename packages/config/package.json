{
  "name": "@ts-gql/config",
  "version": "0.7.1",
  "main": "dist/config.cjs.js",
  "module": "dist/config.esm.js",
  "license": "MIT",
  "dependencies": {
    "@babel/runtime": "^7.9.2",
    "find-pkg-json-field-up": "^1.0.1",
    "lazy-require.macro": "^0.1.0",
    "superstruct": "^0.10.12"
  },
  "peerDependencies": {
    "graphql": "^0.9.0 || ^0.10.0 || ^0.11.0 || ^0.12.0 || ^0.13.0 || ^14.0.0 || ^15.0.0"
  },
  "devDependencies": {
    "graphql": "^14.0.0"
  },
  "repository": "https://github.com/Thinkmill/ts-gql/tree/master/packages/config"
}
