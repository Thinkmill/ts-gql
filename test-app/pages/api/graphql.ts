import { ApolloServer } from "apollo-server-micro";
// @ts-ignore
import typeDefs from "raw-loader!../../schema.graphql";

const resolvers = {
  Query: {
    hello: () => "Hello!",
  },
};

const apolloServer = new ApolloServer({
  typeDefs,
  resolvers,
});

export const config = {
  api: {
    bodyParser: false,
  },
};

export default apolloServer.createHandler({
  path: "/api/graphql",
});
