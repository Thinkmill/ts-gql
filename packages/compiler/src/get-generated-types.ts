import fs from "fs-extra";
import nodePath from "path";
import {
  DocumentNode,
  OperationDefinitionNode,
  parse,
  FragmentDefinitionNode,
  visit,
  validate,
  specifiedRules,
  NoUnusedFragmentsRule,
  GraphQLError,
} from "graphql";
import {
  parse as babelParse,
  transformFromAstAsync,
  PluginObj,
} from "@babel/core";
import { TaggedTemplateExpression } from "@babel/types";
import stripAnsi from "strip-ansi";
import slash from "slash";
import globby from "globby";
import { cachedGenerateSchemaTypes } from "./schema-types";
import { cachedGenerateOperationTypes } from "./operation-types";
import { FsOperation } from "./fs-operations";
import { Config } from "@ts-gql/config";

let fragmentDocumentRules = specifiedRules.filter(
  (x) => x !== NoUnusedFragmentsRule
);

export const getGeneratedTypes = async ({ schema, directory }: Config) => {
  let generatedDirectory = nodePath.join(
    nodePath.join(directory, "__generated__", "ts-gql")
  );

  const files = await globby(["**/*.{ts,tsx}"], {
    cwd: directory,
    absolute: true,
    gitignore: true,
    ignore: ["**/node_modules/**"],
  });

  let nodeMap: Record<
    string,
    {
      filename: string;
      makeFrameError: (error: GraphQLError) => string;
      nodes:
        | readonly [OperationDefinitionNode, ...FragmentDefinitionNode[]]
        | readonly [FragmentDefinitionNode];
    }
  > = {};

  let errors: string[] = [];

  await Promise.all(
    files.map(async (file) => {
      let content = await fs.readFile(file, "utf8");
      if (/gql\s*`/.test(content)) {
        try {
          let ast = babelParse(content, {
            filename: file,
            rootMode: "upward-optional",
          })!;

          await transformFromAstAsync(ast, content, {
            plugins: [
              (): PluginObj => ({
                visitor: {
                  TaggedTemplateExpression(path, state) {
                    let isGqlTag =
                      path.node.tag.type === "Identifier" &&
                      path.node.tag.name === "gql";
                    let hasNoInterpolations =
                      path.node.quasi.quasis.length === 1 &&
                      path.node.quasi.expressions.length === 0;

                    // TODO: verify operation name === import
                    let parentIsAs =
                      path.parent.type === "TSAsExpression" &&
                      path.parent.typeAnnotation.type === "TSImportType";

                    if (isGqlTag && hasNoInterpolations && parentIsAs) {
                      try {
                        let content = path.node.quasi.quasis[0].value.cooked!;
                        let ast = parse(content);

                        let nodes = getGqlNode(ast);
                        let val = nodes[0].name!.value;
                        if (nodeMap[val]) {
                          errors.push(
                            path
                              .buildCodeFrameError(
                                `An operation already exists with the name ${val}`
                              )
                              .toString()
                          );
                        }
                        nodeMap[val] = {
                          filename: slash(nodePath.relative(directory, file)),
                          makeFrameError: (error) => {
                            let loc = locFrom(path.node, error);
                            if (loc) {
                              // @ts-ignore
                              return state.file
                                .buildCodeFrameError(
                                  { loc: { start: loc } },
                                  error.message
                                )
                                .toString();
                            }
                            return path
                              .buildCodeFrameError(error.message)
                              .toString();
                          },
                          nodes,
                        };
                      } catch (err) {
                        errors.push(
                          path.buildCodeFrameError(err.message).toString()
                        );
                      }
                    }
                  },
                },
              }),
            ],
            babelrc: false,
            code: false,
            configFile: false,
            filename: file,
          });
        } catch (err) {
          errors.push(err.toString());
        }
      }
    })
  );

  let fsOperations: FsOperation[] = [];

  try {
    let generatedDirectoryFiles = (await fs.readdir(generatedDirectory))
      .filter((x) => x !== "@schema.d.ts")
      .map((x) => x.replace(/\.ts$/, ""));

    for (let name of generatedDirectoryFiles) {
      if (nodeMap[name] === undefined) {
        fsOperations.push({
          type: "remove",
          filename: nodePath.join(generatedDirectory, name + ".ts"),
        });
      }
    }
  } catch (err) {
    if (err.code !== "ENOENT") {
      throw err;
    }
  }

  let {
    hash: schemaHash,
    operation: schemaOperation,
  } = await cachedGenerateSchemaTypes(schema, generatedDirectory, {});

  if (schemaOperation) {
    fsOperations.push(schemaOperation);
  }

  let dependencies = Object.keys(nodeMap).reduce((obj, item) => {
    obj[item] = [];
    return obj;
  }, {} as Record<string, string[]>);

  for (let key in nodeMap) {
    for (let node of nodeMap[key].nodes) {
      visit(node, {
        FragmentSpread(node) {
          let name = node.name.value;
          if (
            !nodeMap[name] ||
            nodeMap[name].nodes[0].kind === "OperationDefinition"
          ) {
            errors.push(`Fragment ${name} not found`);
          } else {
            dependencies[key].push(name);
          }
        },
      });
    }
  }
  await Promise.all(
    Object.keys(nodeMap).map(async (key) => {
      let nodes = nodeMap[key].nodes.concat(
        getFlatDependenciesForItem(dependencies, key).map(
          (x) => nodeMap[x].nodes[0] as FragmentDefinitionNode
        )
      );
      let document = {
        kind: "Document",
        definitions: nodes,
      } as const;

      let gqlErrors = validate(
        schema,
        document,
        nodes[0].kind === "OperationDefinition"
          ? specifiedRules
          : fragmentDocumentRules
      ).map((err) =>
        // TODO: make this better
        nodeMap[key].makeFrameError(err)
      );

      // TODO: make into proper validation thing
      if (
        nodes[0].kind === "FragmentDefinition" &&
        (!nodes[0].name || !/.+_.+/.test(nodes[0].name.value))
      ) {
        errors.push(
          nodeMap[key].makeFrameError(
            // @ts-ignore
            {
              message:
                "Fragment names must be in the format ComponentName_propName",
            }
          )
        );
      }

      if (gqlErrors.length) {
        errors.push(...gqlErrors);
      }
      let operation = await cachedGenerateOperationTypes(
        schema,
        document,
        nodes[0],
        nodePath.join(generatedDirectory, `${nodes[0].name!.value}.ts`),
        schemaHash,
        nodes[0].name!.value,
        gqlErrors.length
          ? `${nodeMap[key].filename}\nThere ${
              gqlErrors.length === 1 ? "is an error" : "are errors"
            } with ${nodes[0].name!.value}\n${stripAnsi(gqlErrors.join("\n"))}`
          : undefined
      );
      if (operation) fsOperations.push(operation);
    })
  );
  return { fsOperations, errors };
};

function locFrom(node: TaggedTemplateExpression, error: GraphQLError) {
  if (!error.locations || !error.locations.length) {
    return;
  }
  const location = error.locations[0];

  let line;
  let column;
  if (location.line === 1) {
    line = node.loc!.start.line;
    column = node.tag.loc!.end.column + location.column;
  } else {
    line = node.loc!.start.line + location.line - 1;
    column = location.column - 1;
  }

  return {
    line,
    column,
  };
}

function getFlatDependenciesForItem(
  deps: Record<string, string[]>,
  item: string
): string[] {
  return deps[item].concat(
    ...deps[item].map((item) => getFlatDependenciesForItem(deps, item))
  );
}

function getGqlNode(ast: DocumentNode) {
  if (ast.definitions.some((x) => x.kind === "OperationDefinition")) {
    if (
      !ast.definitions.every(
        (x) =>
          x.kind === "OperationDefinition" || x.kind === "FragmentDefinition"
      )
    ) {
      throw new ValidationError(
        "This document has nodes that are not operations or fragments"
      );
    }
    let operationNodes = ast.definitions.filter(
      (x): x is OperationDefinitionNode => x.kind === "OperationDefinition"
    );
    if (operationNodes.length !== 1) {
      throw new ValidationError(
        "This document has more than one operation node"
      );
    }
    let [operationNode] = operationNodes;

    if (!operationNode.name) {
      throw new ValidationError("Operations must have names");
    }
    return ast.definitions as [
      OperationDefinitionNode,
      ...FragmentDefinitionNode[]
    ];
  }
  if (
    ast.definitions.length !== 1 ||
    ast.definitions[0].kind !== "FragmentDefinition"
  ) {
    throw new ValidationError(
      "Documents that do not have an operation must have a single fragment"
    );
  }
  let [fragmentNode] = ast.definitions;
  if (!fragmentNode.name) {
    throw new ValidationError("Fragments must have names");
  }
  return [fragmentNode] as const;
}

class ValidationError extends Error {}
