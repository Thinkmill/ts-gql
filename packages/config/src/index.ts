import path from "path";
import { GraphQLSchema } from "graphql";
import {
  findPkgJsonFieldUp,
  findPkgJsonFieldUpSync,
} from "find-pkg-json-field-up";
import fs from "fs";
import { promisify } from "util";
import { parseSchema, hashSchema } from "./read-schema";

export { parseSchema, hashSchema };

export class ConfigNotFoundError extends Error {}

export type Config = {
  directory: string;
  schema: GraphQLSchema;
  scalars: Record<string, string>;
  addTypename: boolean;
  readonlyTypes: boolean;
  schemaHash: string;
};

export type RawConfig = {
  directory: string;
  schema: string;
  scalars: Record<string, string>;
  addTypename: boolean;
  readonlyTypes: boolean;
};

function parseFieldToConfig({
  packageJson,
  directory,
}: {
  packageJson: Record<string, any>;
  directory: string;
}): RawConfig {
  let field = packageJson["ts-gql"];
  if (
    typeof field === "object" &&
    field !== null &&
    typeof field.schema === "string"
  ) {
    return {
      schema: path.resolve(directory, field.schema),
      directory,
      scalars: field.scalars || {},
      addTypename: field.addTypename ?? true,
      readonlyTypes: field.readonlyTypes ?? true,
    };
  }
  throw new ConfigNotFoundError("ts-gql config not found");
}

export async function getRawConfig(cwd: string) {
  return parseFieldToConfig(await findPkgJsonFieldUp("ts-gql", cwd));
}

export function getRawConfigSync(cwd: string) {
  return parseFieldToConfig(findPkgJsonFieldUpSync("ts-gql", cwd));
}

const readFile = promisify(fs.readFile);

export async function getConfig(cwd: string): Promise<Config> {
  let config = await getRawConfig(cwd);
  let schemaContents = await readFile(config.schema, "utf8");
  return {
    ...config,
    schemaHash: hashSchema(schemaContents),
    schema: parseSchema(config.schema, schemaContents),
  };
}

export function getConfigSync(cwd: string): Config {
  let config = getRawConfigSync(cwd);
  let schemaContents = fs.readFileSync(config.schema, "utf8");
  return {
    ...config,
    schemaHash: hashSchema(schemaContents),
    schema: parseSchema(config.schema, schemaContents),
  };
}
