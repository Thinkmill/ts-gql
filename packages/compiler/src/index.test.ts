import fs from "fs-extra";
import path from "path";
import fixturez from "fixturez";
import { parse } from "graphql";
import { getGeneratedTypes } from "./get-generated-types";
import { getConfig } from "@ts-gql/config";

let f = fixturez(__dirname);

async function build(cwd: string) {
  return getGeneratedTypes(await getConfig(cwd));
}

let gql = ([str]: TemplateStringsArray) => str;

let schema = gql`
  type Query {
    hello: String!
    other: Boolean!
    another: String!
    something: String
    someObj: OutputThing!
    arr: [OutputThing!]!
    thing: Thing!
    optional(thing: String): String!
    oneMore(thing: String, other: Something!): String!
  }

  type Mutation {
    hello: String!
    other: Boolean!
    another: String!
    something: String
    optional(thing: String): String!
    oneMore(thing: String, other: Something!): String!
  }

  interface Thing {
    id: ID!
  }

  type ImplementationOfThing implements Thing {
    id: ID!
    something: String!
  }

  type OutputThing {
    id: ID!
    other: String!
  }

  input Something {
    yes: Boolean
    no: String!
  }
`;

async function setupEnv() {
  let tempdir = f.temp();

  let promises = [];

  promises.push(
    fs.writeFile(
      path.join(tempdir, "package.json"),
      JSON.stringify(
        {
          name: "something",
          "ts-gql": { schema: "schema.graphql" },
        },
        null,
        2
      )
    )
  );
  promises.push(fs.writeFile(path.join(tempdir, "schema.graphql"), schema));
  await Promise.all(promises);
  return tempdir;
}

async function buildAndSnapshot(tempdir: string) {
  let result = await build(tempdir);
  expect({
    errors: result.errors,
    fsOperations: result.fsOperations
      .filter((x) => !path.parse(x.filename).name.startsWith("@"))
      .map((x) => ({ ...x, filename: path.relative(tempdir, x.filename) })),
  }).toMatchSnapshot();
}

function graphql(strs: TemplateStringsArray) {
  let first = parse(strs[0]).definitions[0];
  if (
    first.kind !== "FragmentDefinition" &&
    first.kind !== "OperationDefinition"
  ) {
    throw new Error(
      `First definition is a ${first.kind}, it must be a FragmentDefinition or OperationDefinition`
    );
  }
  if (!first.name) {
    throw new Error("First definition must have a name but it does not");
  }

  return `gql\`${strs[0]}\` as import("./__generated__/ts-gql/${first.name.value}")`;
}

function makeSourceFile(items: string[]) {
  return "import { gql } from '@ts-gql/tag';\n" + items.join("\n");
}

test("basic", async () => {
  let dir = await setupEnv();

  await fs.writeFile(
    path.join(dir, "index.tsx"),
    makeSourceFile([
      graphql`
        query Thing {
          hello
        }
      `,
    ])
  );
  await buildAndSnapshot(dir);
});

test("list with fragment works as expected", async () => {
  let dir = await setupEnv();

  await fs.writeFile(
    path.join(dir, "index.tsx"),
    makeSourceFile([
      graphql`
        query Thing {
          ...Frag_a
          ...Frag_b
        }
      `,
      graphql`
        fragment Frag_a on Query {
          arr {
            id
          }
        }
      `,
      graphql`
        fragment Frag_b on Query {
          arr {
            other
          }
        }
      `,
    ])
  );

  await buildAndSnapshot(dir);
});
