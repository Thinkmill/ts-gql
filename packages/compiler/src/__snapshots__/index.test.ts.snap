// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic 1`] = `
Object {
  "errors": Array [],
  "fsOperations": Array [
    Object {
      "content": "// ts-gql-integrity:c81cc439d4fcdfdab206fbcfb8b23920
/*
ts-gql-meta-begin
{
  \\"hash\\": \\"b1e504b028fd1d4b0a420bf35ebeee49\\"
}
ts-gql-meta-end
*/

import * as SchemaTypes from \\"./@schema\\";
import { TypedDocumentNode } from \\"@ts-gql/tag\\";

type ThingQueryVariables = {};


type ThingQuery = (
  { readonly __typename: 'Query' }
  & Pick<SchemaTypes.Query, 'hello'>
);


export type type = TypedDocumentNode<{
  type: \\"query\\";
  result: ThingQuery;
  variables: ThingQueryVariables;
  documents: SchemaTypes.TSGQLDocuments;
}>

declare module \\"./@schema\\" {
  interface TSGQLDocuments {
    Thing: type;
  }
}

export const document = JSON.parse(\\"{\\\\\\"kind\\\\\\":\\\\\\"Document\\\\\\",\\\\\\"definitions\\\\\\":[{\\\\\\"kind\\\\\\":\\\\\\"OperationDefinition\\\\\\",\\\\\\"operation\\\\\\":\\\\\\"query\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"Thing\\\\\\"},\\\\\\"variableDefinitions\\\\\\":[],\\\\\\"directives\\\\\\":[],\\\\\\"selectionSet\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"SelectionSet\\\\\\",\\\\\\"selections\\\\\\":[{\\\\\\"kind\\\\\\":\\\\\\"Field\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"hello\\\\\\"},\\\\\\"arguments\\\\\\":[],\\\\\\"directives\\\\\\":[]}]}}]}\\")
",
      "filename": "__generated__/ts-gql/Thing.ts",
      "type": "output",
    },
  ],
}
`;

exports[`list with fragment works as expected 1`] = `
Object {
  "errors": Array [],
  "fsOperations": Array [
    Object {
      "content": "// ts-gql-integrity:ccdc658c85c75263d36976479f4578c1
/*
ts-gql-meta-begin
{
  \\"hash\\": \\"fe6dbfdcf052e52348034ef3f70d70a4\\"
}
ts-gql-meta-end
*/

import * as SchemaTypes from \\"./@schema\\";
import { TypedDocumentNode } from \\"@ts-gql/tag\\";

type ThingQueryVariables = {};


type ThingQuery = (
  { readonly __typename: 'Query' }
  & { readonly arr: ReadonlyArray<(
    { readonly __typename: 'OutputThing' }
    & Pick<SchemaTypes.OutputThing, 'id' | 'other'>
  )> }
);


export type type = TypedDocumentNode<{
  type: \\"query\\";
  result: ThingQuery;
  variables: ThingQueryVariables;
  documents: SchemaTypes.TSGQLDocuments;
}>

declare module \\"./@schema\\" {
  interface TSGQLDocuments {
    Thing: type;
  }
}

export const document = JSON.parse(\\"{\\\\\\"kind\\\\\\":\\\\\\"Document\\\\\\",\\\\\\"definitions\\\\\\":[{\\\\\\"kind\\\\\\":\\\\\\"OperationDefinition\\\\\\",\\\\\\"operation\\\\\\":\\\\\\"query\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"Thing\\\\\\"},\\\\\\"variableDefinitions\\\\\\":[],\\\\\\"directives\\\\\\":[],\\\\\\"selectionSet\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"SelectionSet\\\\\\",\\\\\\"selections\\\\\\":[{\\\\\\"kind\\\\\\":\\\\\\"FragmentSpread\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"Frag_a\\\\\\"},\\\\\\"directives\\\\\\":[]},{\\\\\\"kind\\\\\\":\\\\\\"FragmentSpread\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"Frag_b\\\\\\"},\\\\\\"directives\\\\\\":[]}]}},{\\\\\\"kind\\\\\\":\\\\\\"FragmentDefinition\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"Frag_a\\\\\\"},\\\\\\"typeCondition\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"NamedType\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"Query\\\\\\"}},\\\\\\"directives\\\\\\":[],\\\\\\"selectionSet\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"SelectionSet\\\\\\",\\\\\\"selections\\\\\\":[{\\\\\\"kind\\\\\\":\\\\\\"Field\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"arr\\\\\\"},\\\\\\"arguments\\\\\\":[],\\\\\\"directives\\\\\\":[],\\\\\\"selectionSet\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"SelectionSet\\\\\\",\\\\\\"selections\\\\\\":[{\\\\\\"kind\\\\\\":\\\\\\"Field\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"id\\\\\\"},\\\\\\"arguments\\\\\\":[],\\\\\\"directives\\\\\\":[]}]}}]}},{\\\\\\"kind\\\\\\":\\\\\\"FragmentDefinition\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"Frag_b\\\\\\"},\\\\\\"typeCondition\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"NamedType\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"Query\\\\\\"}},\\\\\\"directives\\\\\\":[],\\\\\\"selectionSet\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"SelectionSet\\\\\\",\\\\\\"selections\\\\\\":[{\\\\\\"kind\\\\\\":\\\\\\"Field\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"arr\\\\\\"},\\\\\\"arguments\\\\\\":[],\\\\\\"directives\\\\\\":[],\\\\\\"selectionSet\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"SelectionSet\\\\\\",\\\\\\"selections\\\\\\":[{\\\\\\"kind\\\\\\":\\\\\\"Field\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"other\\\\\\"},\\\\\\"arguments\\\\\\":[],\\\\\\"directives\\\\\\":[]}]}}]}}]}\\")
",
      "filename": "__generated__/ts-gql/Thing.ts",
      "type": "output",
    },
    Object {
      "content": "// ts-gql-integrity:a2441203eeb49e58a2cfa70fc14f58be
/*
ts-gql-meta-begin
{
  \\"hash\\": \\"b069003deb9f8d640946a54d9dc22c59\\"
}
ts-gql-meta-end
*/

import * as SchemaTypes from \\"./@schema\\";
import { TypedDocumentNode } from \\"@ts-gql/tag\\";

type Frag_aFragment = (
  { readonly __typename: 'Query' }
  & { readonly arr: ReadonlyArray<(
    { readonly __typename: 'OutputThing' }
    & Pick<SchemaTypes.OutputThing, 'id'>
  )> }
);


export type type = TypedDocumentNode<{
  type: \\"fragment\\";
  result: Frag_aFragment;
  documents: SchemaTypes.TSGQLDocuments;
}>

declare module \\"./@schema\\" {
  interface TSGQLDocuments {
    Frag_a: type;
  }
}

export const document = JSON.parse(\\"{\\\\\\"kind\\\\\\":\\\\\\"Document\\\\\\",\\\\\\"definitions\\\\\\":[{\\\\\\"kind\\\\\\":\\\\\\"FragmentDefinition\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"Frag_a\\\\\\"},\\\\\\"typeCondition\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"NamedType\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"Query\\\\\\"}},\\\\\\"directives\\\\\\":[],\\\\\\"selectionSet\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"SelectionSet\\\\\\",\\\\\\"selections\\\\\\":[{\\\\\\"kind\\\\\\":\\\\\\"Field\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"arr\\\\\\"},\\\\\\"arguments\\\\\\":[],\\\\\\"directives\\\\\\":[],\\\\\\"selectionSet\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"SelectionSet\\\\\\",\\\\\\"selections\\\\\\":[{\\\\\\"kind\\\\\\":\\\\\\"Field\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"id\\\\\\"},\\\\\\"arguments\\\\\\":[],\\\\\\"directives\\\\\\":[]}]}}]}}]}\\")
",
      "filename": "__generated__/ts-gql/Frag_a.ts",
      "type": "output",
    },
    Object {
      "content": "// ts-gql-integrity:6209af8a099912a7d3ca0cd547392d2d
/*
ts-gql-meta-begin
{
  \\"hash\\": \\"287f04e0ea48ed59cad93b99c244385c\\"
}
ts-gql-meta-end
*/

import * as SchemaTypes from \\"./@schema\\";
import { TypedDocumentNode } from \\"@ts-gql/tag\\";

type Frag_bFragment = (
  { readonly __typename: 'Query' }
  & { readonly arr: ReadonlyArray<(
    { readonly __typename: 'OutputThing' }
    & Pick<SchemaTypes.OutputThing, 'other'>
  )> }
);


export type type = TypedDocumentNode<{
  type: \\"fragment\\";
  result: Frag_bFragment;
  documents: SchemaTypes.TSGQLDocuments;
}>

declare module \\"./@schema\\" {
  interface TSGQLDocuments {
    Frag_b: type;
  }
}

export const document = JSON.parse(\\"{\\\\\\"kind\\\\\\":\\\\\\"Document\\\\\\",\\\\\\"definitions\\\\\\":[{\\\\\\"kind\\\\\\":\\\\\\"FragmentDefinition\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"Frag_b\\\\\\"},\\\\\\"typeCondition\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"NamedType\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"Query\\\\\\"}},\\\\\\"directives\\\\\\":[],\\\\\\"selectionSet\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"SelectionSet\\\\\\",\\\\\\"selections\\\\\\":[{\\\\\\"kind\\\\\\":\\\\\\"Field\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"arr\\\\\\"},\\\\\\"arguments\\\\\\":[],\\\\\\"directives\\\\\\":[],\\\\\\"selectionSet\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"SelectionSet\\\\\\",\\\\\\"selections\\\\\\":[{\\\\\\"kind\\\\\\":\\\\\\"Field\\\\\\",\\\\\\"name\\\\\\":{\\\\\\"kind\\\\\\":\\\\\\"Name\\\\\\",\\\\\\"value\\\\\\":\\\\\\"other\\\\\\"},\\\\\\"arguments\\\\\\":[],\\\\\\"directives\\\\\\":[]}]}}]}}]}\\")
",
      "filename": "__generated__/ts-gql/Frag_b.ts",
      "type": "output",
    },
  ],
}
`;
